### PR TITLE
Restore os.path methods overload workaround

### PR DESCRIPTION
**Pull request by Avasam**
See https://github.com/python/mypy/issues/17952 and https://github.com/pypa/setuptools/pull/4688/files#diff-e3d3d454fa3a072c9f46f8affa27513892fbc2d245e87f57a5927a7be851de05R2651
Reverts https://github.com/python/typeshed/pull/12208 and adds tests

CC @hauntsaninja @srittau

### CODE REVIEW

**Review by hauntsaninja — COMMENTED** [review]
Nice, thanks! Let's remove the references to https://github.com/python/mypy/issues/3644 since that issue is fixed and replace them with reference to https://github.com/python/mypy/issues/17952 which we can use to track the remaining issue with Any.

(We could also link the issue in re, as Alex mentions here https://github.com/python/typeshed/pull/12208#pullrequestreview-2138452109 )

**Comment by Avasam** [comment]
> (We could also link the issue in re, as Alex mentions here [#12208 (review)](https://github.com/python/typeshed/pull/12208#pullrequestreview-2138452109) )

I *think* I did what you meant here. But I'm not certain.

File: stdlib/posixpath.pyi [hunk 1]
```diff
@@ -77,7 +77,11 @@ pathsep: LiteralString
 defpath: LiteralString
 devnull: LiteralString
 
-def abspath(path: PathLike[AnyStr] | AnyStr) -> AnyStr: ...
+# Overloads were necessary to work around python/mypy#17952 & python/mypy#11880
```
Comment (Samuel-Therrien-Beslogic) [inline stdlib/posixpath.pyi:80]:
```suggestion
# Overloads are necessary to work around python/mypy#17952 & python/mypy#11880
```

**Review by srittau — APPROVED** [review]

**Review by hauntsaninja — APPROVED** [review]
Nice, thank you! Hopefully these get fixed and we can reland the cleanup :-)

**Pull request merged**